fableous-fe/src/components/Canvas.tsx

```diff
@@ -18,417 +18,609 @@ import { ControllerRole, ToolMode, WSMessage, WSMessageType } from "../Data";
 const ASPECT_RATIO = 9 / 16;
 const SCALE = 2;
 
-const Canvas = (props: {
+const SELECT_PADDING = 8;
+
+interface Shape {
+  x1: number;
+  y1: number;
+  x2: number;
+  y2: number;
+  selected: boolean;
+}
+
+interface TextShape extends Shape {
+  text: string;
+  fontSize: number;
+}
+
+interface CanvasProps {
   wsRef: MutableRefObject<WebSocket | undefined>;
   role: ControllerRole;
   layer: ControllerRole;
-}) => {
-  const { layer, role, wsRef } = props;
-  const canvasRef = useRef<HTMLCanvasElement>(document.createElement("canvas"));
-  const [allowDrawing, setAllowDrawing] = useState(false);
-  const [drawing, setDrawing] = useState(false);
-  const [lastPos, setLastPos] = useState([0, 0]);
-  const [toolColor, setToolColor] = useState("#000000ff");
-  const [toolMode, setToolMode] = useState<ToolMode>(ToolMode.None);
-  const [toolWidth, setToolWidth] = useState(8 * SCALE);
+}
 
-  const translateXY = (x: number, y: number) => {
-    const bound = canvasRef.current.getBoundingClientRect();
-    return [(x - bound.x) * SCALE, (y - bound.y) * SCALE];
-  };
-
-  const scaleDownXY = (x: number, y: number) => {
-    return [x / canvasRef.current.width, y / canvasRef.current.height];
-  };
+const Canvas = forwardRef<HTMLCanvasElement, CanvasProps>(
+  (props: CanvasProps, ref) => {
+    const { layer, role, wsRef } = props;
+    const canvasRef = ref as MutableRefObject<HTMLCanvasElement>;
+    const [allowDrawing, setAllowDrawing] = useState(false);
+    const [dragging, setDragging] = useState(false);
+    const [hasLifted, setHasLifted] = useState(false);
+    const [editingText, setEditingText] = useState(0);
+    const [lastPos, setLastPos] = useState([0, 0]);
+    const [textId, setTextId] = useState(1);
+    const [textShapes, setTextShapes] = useState<{ [id: string]: TextShape }>(
+      {}
+    );
+    const [toolColor, setToolColor] = useState("#000000ff");
+    const [toolMode, setToolMode] = useState<ToolMode>(ToolMode.None);
+    const [toolWidth, setToolWidth] = useState(8 * SCALE);
 
-  const scaleUpXY = (x: number, y: number) => {
-    return [x * canvasRef.current.width, y * canvasRef.current.height];
-  };
+    const translateXY = (x: number, y: number) => {
+      const bound = canvasRef.current.getBoundingClientRect();
+      return [(x - bound.x) * SCALE, (y - bound.y) * SCALE];
+    };
 
-  const convHEXtoRGBA = (hex: string) => {
-    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(
-      hex
+    const scaleDownXY = useCallback(
```

Comment:
after a bit of reading, is it ture that this func is using useCallback() over just vanilla func because its used inside a useEffect() hook? if so this is not apparent during skimming, maybe add a short comment? There are multiple occurences, but just here should be enough

Reply:
correct, got it